### DISCUSSION TITLE
How to run kanel in an app using a forked version of this repo?

### DISCUSSION THREAD

**Discussion by christiaanwesterbeek** [kristiandupont/kanel#371] [General]
I have made a [fork](https://github.com/christiaanwesterbeek/kanel/tree/additional-views-support) from this repo with some minor changes that will also try to look for a target in the views. 

The problem is that when I run `yarn exec kanel` that it runs the version of https://github.com/kristiandupont/kanel.

Let's dive in.

I've added my fork to some repo like this.

`yarn add -D https://github.com/christiaanwesterbeek/kanel#additional-views-support`

In the dependencies in node_modules, I can actually see my changes here
<img width="879" alt="image" src="https://user-images.githubusercontent.com/465989/217864615-1b1d768d-331f-4689-aaee-32960b5d8521.png">

But not here
<img width="1070" alt="image" src="https://user-images.githubusercontent.com/465989/217864787-3a039635-8612-4d76-b99c-bb883c6e7f32.png">

So, my question is how can I add my fork of kanel to some repo and execute kanel using my forks code, instead of kristiandupont's?

**Comment by kristiandupont** [comment]
I might be missing something but it looks like your version wasn't built? Did you run `yarn build` inside `/packages/kanel`?

**Comment by christiaanwesterbeek** [comment]
When in my local clone of my fork of kanel, I run `yarn build` inside /packages/kanel, I indeed see my changes in /packages/kanel/src/generators/resolveType.ts are built in /packages/kanel/build/generators/resolveType.js. The thing is that that directory is ignored, so not committed.

Now back to my application, where I have my fork as a dependency. This is the package.json of that application

```
...
"devDependencies": {
    "dotenv": "^16.0.3",
    "eslint-plugin-ignore-generated": "^3.1.0",
    "kanel-monorepo": "https://github.com/christiaanwesterbeek/kanel#additional-views-support"
}
```

The 2 screenshots in my first comment are from that application. As you can see, in the build directory there is not what would be the build result of my code. It seems unlikely that I need to go inside of node_modules/kanel-monorepo/packages/kanel to run `yarn build` there.

The problem I'm having should be easy to replicate. In any app where you already have kanel as dependancy, try to replace that with my fork, do yarn install, and see what's inside node_modules/kanel/build/generators/resolveType.js.

**Comment by kristiandupont** [comment]
Looking at your screenshots, you see your changes in a `src` folder that sits right underneath `bin`. In the second screenshot, the folder right underneath `bin` is the `build` folder. Are you referring to the same `node_modules` folder?

**Comment by kristiandupont** [comment]
Right, I think I understood that the second screenshot is from your clone, and the first one is from where you installed it?

**Comment by kristiandupont** [comment]
What does
```bash
ls -la node_modules/.bin/kanel
```
give you?

**Comment by christiaanwesterbeek** [comment]
```
➜  myapp> ls -la node_modules/.bin/kanel
lrwxr-xr-x  1 chris  staff  18 Feb  9 17:42 node_modules/.bin/kanel -> ../kanel/bin/kanel
```

Both screenshots are from the node_modules directory in my application. Just scrolled a bit and expanded the different folders. Just to clarify, here selected the 2 folders that I made the screenshots off:
<img width="217" alt="image" src="https://user-images.githubusercontent.com/465989/217917437-5cf01960-442d-4e3d-bb01-4ce298ba43b6.png">

**Comment by kristiandupont** [comment]
Oh, so you have both the regular Kanel installed as well as your fork?

**Comment by christiaanwesterbeek** [comment]
I don't think so, because this is the only thing I installed:

`yarn add -D https://github.com/christiaanwesterbeek/kanel#additional-views-support`

And the earlier package.json I showed and the screenshots I showed are the result of just that...

**Comment by christiaanwesterbeek** [comment]
I'm going to try that again, with a fresh blank app.

**Comment by christiaanwesterbeek** [comment]
```
mkdir myapp
cd myapp
yarn init
yarn add -D https://github.com/christiaanwesterbeek/kanel\#additional-views-support
```
Both kanel and kanel-monorepo in node_modules
<img width="245" alt="image" src="https://user-images.githubusercontent.com/465989/217919206-9d1792c0-6807-42ff-a0a8-e47cb15a1ef4.png">

kanel-monorepo has my code
<img width="758" alt="image" src="https://user-images.githubusercontent.com/465989/217919487-ab3aa5bb-bbe4-46da-9df5-f1653d82e665.png">

kanel does not have my code
<img width="889" alt="image" src="https://user-images.githubusercontent.com/465989/217919597-dd3b3e83-e684-490f-88a8-e39dc65a6eab.png">

**Comment by kristiandupont** [comment]
No. I suggest you `yarn remove kanel` first so it doesn't interfere. I don't know what it will do if two packages try to create the same file in .bin but I suppose it's just a last-one-wins.

**Reply by christiaanwesterbeek** [reply]
You mean I should do?

```
mkdir myapp
cd myapp
yarn init
yarn add -D https://github.com/christiaanwesterbeek/kanel#additional-views-support
yarn remove kanel
```

That seems weird. I just reproduced the problem in a blank app, there was no kanel in the first place in there...

(or did we cross-post?)

**Comment by kristiandupont** [comment]
Ah of course, it's adding it because the mono repo itself refers to kanel (https://github.com/kristiandupont/kanel/blob/master/package.json#L32) -- maybe remove that in your fork. But you will need to manually create a symbolic link to the built version to run the cli.

**Reply by christiaanwesterbeek** [reply]
That is probably it. Thank you!!

I'm going to try that. Is the symbolic link something that I should commit in my fork?

**Comment by kristiandupont** [comment]
Well the thing is that you are not installing Kanel but the mono-repo the way you are doing it. The mono-repo doesn't have any CLI or anything, nor should it. The package itself is just in the `packages/kanel` folder, which is what is deployed to NPM.

**Reply by christiaanwesterbeek** [reply]
What would be an easier way to run a forked version of kanel in my app? Because I think I started off on the wrong foot by doing `yarn add -D https://github.com/christiaanwesterbeek/kanel#additional-views-support`

**Reply by kristiandupont** [reply]
Good question!

I tend to just run it where I have it checked out (specifically, I tend to modify the .kanelrc file in the example folder) and inspect the results. Of course, if you need to compile the resulting code together with the rest of your code base, that won't work. You *could* replace the kanel folder in node_modules with a symlink to your `kanel-monorepo/packages/kanel` folder. It's a bit of a hassle but it should work.

**Reply by kristiandupont** [reply]
Actually, now that I think of it, I guess changing the symlink in `node_modules/.bin` to whereever you have it checked out should be sufficient. But I haven't tried it.

**Reply by christiaanwesterbeek** [reply]
That worked! I now simply cloned the kanel repo adjacent to my app. I checked out my feature branch, did `yarn install`, then `cd ./packages/kanel`, then `yarn build`.

Then I went back to my app and opened `node_modules/.bin/kanel` and modified it from this
```
#!/usr/bin/env node
require('../build/cli.js');
```

into this:
```
#!/usr/bin/env node
require('../../../../kanel/packages/kanel/build/cli.js');
```

Now In my app when I run `yarn exec kanel`, then the build of the adjacent forked kanel app is used!

Thanks for helping me out @kristiandupont !

**Reply by kristiandupont** [reply]
Awesome! I should probably add this to the README, it seems like good developer info!
